Feature/57 feature backend architektur refactoring ports adapters

NicKIT01 opened the pull request:
You are welcome to give feedback and suggestions.

NicKIT01 committed:
#57: Preparing for Hexagonal Architecture: transform Godclass models into domain POJOs

NicKIT01 committed:
#57: Small adjustments in POJO files

NicKIT01 committed:
#57: Created Jakarta Persistence API (JPA) files and duplicated configurations in infrastructure

NicKIT01 committed:
#57: Definition of ports:
- outgoing: repository interfaces representing database contracts
- incoming: dispatcher which defines use cases for CRUD operations

NicKIT01 committed:
#57: remove unused imports

NicKIT01 committed:
#57: introduce bidirectional mapping between POJOs and Jpa files using MapStruct

NicKIT01 committed:
#57: introduce filesystem structure in infrastructure and core

NicKIT01 committed:
#57: remove import error

NicKIT01 committed:
#57: Spring data repositories for CRUD Operations

NicKIT01 committed:
#57: refine outgoing ports and implement JPA based adapters to decouple domain from persistence. This replaces the old FileElementSaver approach

NicKIT01 committed:
#57: add FileSystemUseCase incoming port for application layer defining all incoming operations for the filesystem domain

NicKIT01 committed:
#57: single SpringDataJpaFileElementRepository for FileElement

NicKIT01 committed:
#57: redifining outgoing ports

NicKIT01 committed:
#57: adjust Jpa adapters to outgoing port structure

NicKIT01 committed:
#57: speficify service ports and services

NicKIT01 committed:
#57: api rest controllers

NicKIT01 committed:
#57: move app to quak/app

NicKIT01 committed:
#57: small adjustments in domain logic

NicKIT01 committed:
#57: Refactor filesystem domain into clean hexagonal architecture
- Updated package structure across the entire module => created and refactored current structure to clean ports & adapters structure divided in core/, infrastructure/, application/
- Adjusted/Introduce incoming and outgoing ports: application/filesystem/ports (incoming/outgoing)
- Added application-level services for filesystem operations
- Separated infrastructure adapters (web + JPA)
- Introduced DTOs and mappers for Directory, File, Project
- Removed old polymorph FileElement-based API responses => 1 service and api for each directory, file, project
- Updated/Introduced controllers to use new DTO/Mapper stack

- Added TODO markers for incomplete logic

This is a structural refactoring preparing for further work on the new architecture and does not yet represent a complete implementation of the previous functionality.

NicKIT01 committed:
Merge branch 'development' into feature/57-feature-backend-architektur-refactoring-ports-adapters

NicKIT01 committed:
#57: Package structure/naming updates for consistent handling of adapters and mappers

NicKIT01 committed:
#57: Introduced FileContent Table and only retrieve FileContent not whole file on request

NicKIT01 committed:
#57: Introduce Container Delegator pattern to select the right FileElementContainer repository at runtime

NicKIT01 committed:
#57: FileService handles use cases related to File. File is persisted through Aggregate root (Directory|Project) and FileContentRepository. FileRepository is read only.

Jooshua commented on backend/src/main/java/edu/kit/quak/application/filesystem/ports/outgoing/ProjectRepositoryPort.java:
```diff
@@ -0,0 +1,29 @@
+package edu.kit.quak.application.filesystem.ports.outgoing;
```
What is the difference between `ports.out` and `ports.outgoing`?

NicKIT01 replied:
Einfach nur konsistentes naming. Wir haben uns darauf geeinigt alles mit out und in zu betiteln Hält die package Struktur kürzer und sagt trotzdem noch aus, was im package ist.

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/dto/RenameFileRequest.java:
```diff
@@ -0,0 +1,8 @@
+import jakarta.validation.constraints.Size;
+
+public record RenameFileRequest(
+        @Size(min = 1, message = "Name must not be empty.")
```
The other DTOs use `@NotBlank`, why not also use it here?

Jooshua reviewed (commented):
I had a short look into it and I really like the new structure :) I think having an ADR will be really helpful in the future.

NicKIT01 committed:
#57:
- DirectoryService handles use cases related to Directory. Directory is persisted through Aggregate root (Directory|Project).
- Change rename by renaming through parent

NicKIT01 committed:
#57:
- ProjectService handles use cases related to Project. Project is always aggregate root

NicKIT01 committed:
#57: Delete CustomIDGenerator and use UUID instead

NicKIT01 committed:
#57: Delete FileElementResolver.java and delete remaining generateId methods in JPA Entities

NicKIT01 committed:
#57: Consistent validation of DTOs

NicKIT01 committed:
#57: Correct the usage of FileElementDto

NicKIT01 committed:
#57: Refined JpaMapper and JpaEntities
1. Correct polymorphic mapping between FileElement<?> and JpaFileElement<?> and between Sets.
2. Seperation of Concerns: introduced concrete JpaMapper
3. Delete Jackson annotations and Configurations we use DTOs instead

NicKIT01 committed:
#57: Delete old files architecture

NicKIT01 committed:
#57: Domain Unit Tests
1. Delete old files tests
2. Introduce Domain Tests
3. Fix arising errors in domain

NicKIT01 committed:
#57: JpaMapper Unit Tests with Mockito

NicKIT01 committed:
#57: Jpa Adapter Tests:
1. Created ProjectJpaAdapterTest
2. Fix setting parent after mapping
3. Fix equals and hashCode methods which violated Contract (Not the same hashCode and danger of StackOverflow if child and parent call each others hash)

NicKIT01 committed:
#57: implement DirectoryJpaAdapterTests

NicKIT01 committed:
#57: File related JpaAdapterTests
1. Create FileJpaAdapterTest
2. Manually attach parent to File after mapping using shallow methods
3. Create FileContentJpaAdapterTest

NicKIT01 committed:
#57: Add test tag

NicKIT01 committed:
#57: Add domain prefix to separate Project, Directory, File by id

NicKIT01 committed:
#57: Change Adapter id handling to domain prefixes and test behavior

NicKIT01 committed:
#57: Test Services with Mockito

NicKIT01 committed:
#57: Delegator Tests with dummy classes

NicKIT01 committed:
#57: FileElementDtoMapperTest
1. Write Test
2. Resolve cyclic dependency with Lazy loading

NicKIT01 committed:
#57: GlobalExceptionHandler
1. Introduce Global Exception Handler via RestControllerAdvice and ProblemDetails (RFC-7807)
2. Remove error handling in Controller
3. Adjust Service Ports and Services

NicKIT01 committed:
#57: Bugfix: GlobalExceptionHandler - remove ResponseEntityExceptionHandler dependency

NicKIT01 committed:
#57: Create WebMVCTests for Rest Adapters

NicKIT01 committed:
#57: Within a container, the name is now unique.

NicKIT01 committed:
#57: Integration Test for creating a project, directory and file

NicKIT01 committed:
#57: Load parent from database not from child

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Delete bidirectional parent object in child objects (domain)
2. Add a reference by parentId (domain)
3. Domain Tests

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Adjust use cases to use parentId instead of parent
2. Test services

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Adjust and clean up Jpa Entity models

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Simplify mappers

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Set parent mapping in Adapters
2. Adjust Adapter tests

NicKIT01 committed:
#57: Refactoring: Referencing Aggregates by Identity
1. Adjust Dtos to contain LastAccess and CreatedOn for every FileElement

NicKIT01 committed:
#57: Extended Integration Test

NicKIT01 committed:
#57: Architecture description

NicKIT01 committed:
Merge branch 'development' into feature/57-feature-backend-architektur-refactoring-ports-adapters

NicKIT01 committed:
#57: Delete TestUtil and use Inject Mocks instead

NicKIT01 committed:
#57: Change frontend accordingly to new backend Structure

NicKIT01 committed:
#57: Bugfix: encode FileContent before sending it

Jooshua commented on backend/src/main/java/edu/kit/quak/application/filesystem/delegator/FileElementContainerRepositoryDelegator.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Map;
+import java.util.Optional;
+
+@Component
```
Ich glaube hier wäre ein JavaDoc noch ganz gut. Also wir brauchen die Klasse da wir verschiedene Repositories haben über die Generics (file, directory, project) und die klasse macht den lookup von type / prefix zu repository und delegiert dann entsprechend `save` und `find`, oder? 

Vielleicht auch noch erklären warum wir `save` über den type und `find` über den prefix machen. 

Macht es Sinn hier die Lookup Funktionalität aus der Klasse rauszuziehen in ne separate `Mapper` Klasse?

NicKIT01 replied:
Mit Kommentaren war ich echt etwas sparsam. Eigentlich hast du es im Kern schon richtig erfasst. Wir brauchen die im Kern aus dem Grund, dass wir in DDD über Aggregate roots persistieren und nicht mehr über die Leaves selbst. Ich habe das etwas aufgebrochen um performanter zu sein (kleinere Objektbäume) und speichere über den konkreten Aggregator/parent. 

Bei der save Methode haben wir den parent ja schon und können deshalb direkt den Typ prüfen, beim findById haben wir diesen nicht, sondern nur die Id des parents, weswegen wir über einen prefix gehen müssen. Theoretisch könnte man bei der save auch über den prefix gehen, hat sich für mich so aber irgendwie Typsicherer angefühlt. 

Das Auslagern macht wahrscheinlich echt Sinn vor allem in Bezug auf Seperation of Concerns und Testbarkeit. Ich würde aber nicht in einen Mapper auslagern, sondern in eine lookup Registry o.ä. Gehe da nochmal drüber, vielen Dank für die Idee und die Anmerkung :)

NicKIT01 replied:
Hab das nochmal refactored und in eine registry ausgelagert. Ich gehe nur noch über prefixes jetzt, da ich den parent sowieso erst über findById lade.

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/DirectoryRestAdapter.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PostMapping("/")
+    @ResponseStatus(HttpStatus.CREATED)
+    public DirectoryDetailsResponse createDirectory(@RequestBody DirectoryRequest request, @RequestHeader(name = "parent-id") String parentId) {
```
`"parent-id"` ist an mehreren Stellen in Adaptern und Tests hard coded

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/ProjectRestAdapter.java:
```diff
@@ -0,0 +1,56 @@
+package edu.kit.quak.infrastructure.filesystem.in.web.rest;
```
Hier könnte man dann noch einmal schauen ob die OpenAPI Spezifikation changes hat.

NicKIT01 replied:
Ja das stimmt, ich dachte ich warte auf die automatische Generierung, auch weil ich noch ein Issue zu Exceptions und Logging geschrieben habe, dan wird sich da ja eh nochmal was ändern

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/advice/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,63 @@
+        return problem;
+    }
+
+    // Catches "Corrupt State" amd Configuration errors -> 500 Internal Server Error
```
typo

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,63 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    // Catches Validation errors -> 400 Bad Request
```
Passiert die Validation nur bei den `@Valid` Annotationen, oder gibt es noch andere Trigger? Ggf. hier noch einmal spezifizieren.

NicKIT01 replied:
Würde ich im Exception Refactoring angehen, habe auch Validations in den Domains. Ich glaube ein Mix ist sinnvoll, bin mir da aber auch noch nicht 100 prozentig sicher.

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,63 @@
+    // Fallback -> 500 Internal Server Error
+    @ExceptionHandler(Exception.class)
+    public ProblemDetail handleGeneralError(Exception ex) {
+        // TODO: Introduce Logging! (Log.error(ex))
```
Ja :thumbsup:

Jooshua commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/mapper/FileElementDtoMapper.java:
```diff
@@ -0,0 +1,46 @@
+
+    public FileElementDto toDto(FileElement<?> element) {
+        if (element instanceof File file) {
+            return fileMapper.toDetailsResponse(file); // oder @Autowired Mapper, wenn Spring
```
was meinst du mit `// oder @Autowired Mapper, wenn Spring`?

NicKIT01 replied:
Gute Frage, ich glaub das ist ein Relikt aus der Implementierung. Ich injiziere ja mittlerweile über Spring mit Autowired😅

Jooshua commented on backend/ARCHITECTURE.md:
```diff
@@ -0,0 +1,28 @@
+  </tr>
+</table>
+
+Hexagonal Architecture solves this by defining [`incoming & outgoing ports`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/application/filesystem/ports) via dependency injection.
+
+[`Input ports`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/application/filesystem/ports/in) specify the interface through which the application can be accessed. 
+Various adapters—such as [`web controllers`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/infrastructure/in/web/rest), messaging systems, or other external systems—use these input ports to invoke the use cases via dependency injection.
+
+The [`use cases`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/application/filesystem/services) implement the input ports and encapsulate the **business rules**, operating on [`domain entities`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/core/filesystem/model) that represent the core domain model. 
+
+[`Output ports`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/application/filesystem/ports/out) define how the system interacts with external resources, such as databases or external services.
+
+[`Persistence Adapters`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/infrastructure/out/db/jpa) implement these [`output ports`](https://github.com/MoQel/QuaK/tree/development/backend/src/main/java/edu/kit/quak/application/filesystem/ports/out), allowing the infrastructure to be swapped or replaced without affecting the core business logic, ensuring high main ...
```
Finde ich sehr gut. Vielleicht könntest du jede Ebene noch einmal in einem Detailgrad mehr beschreiben. Also z.B. bei `rest`: Hier haben wir in der Regel `RestAdapter`, die machen ...; `DTOs`, ...; `DtoMapper`, ...

Jooshua commented on backend/src/main/java/edu/kit/quak/files/GateController.java:
```diff
@@ -1,34 +0,0 @@
-package edu.kit.quak.files;
```
Die Gates brauchen wir nicht mehr / kommen dann in einem anderen Ticket?

NicKIT01 replied:
Good point, das muss ich nochmal überprüfen

NicKIT01 replied:
Hab das voll übersehen, gerade weil es in der alten Architektur unter files lag. Und die Änderung aus einem der merges von Development in meinen feature branch kam. Hat sich also eingeschlichen. Habs jetzt trotzdem mal komplett wieder integriert, refactored und getestet. Funktioniert jetzt einwandfrei :)

Jooshua replied:
Ggf. noch mit @manuelcorcoles crosschecken?

Jooshua commented on backend/src/main/java/edu/kit/quak/files/configuration/DepthFilter.java:
```diff
@@ -1,23 +0,0 @@
-package edu.kit.quak.files.configuration;
```
Laden wir jetzt immer den kompletten Datei-Baum?

NicKIT01 replied:
Nein, wir haben ein Mapping auf unsere Dtos und auf die Domain Objekte, dabei wird immer nur flach gemappt und es wird maximal 1 Layer an children mit nach außen geliefert. Diese Filter waren vorher notwendig, da wir unsere Domain nach außen getragen hatten. Wir hatten als return type der Rest Controller z.B. FileElement<?>. Sind ja jetzt Dtos.

Jooshua commented on backend/src/test/java/edu/kit/quak/application/filesystem/services/DirectoryServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+    private DirectoryService service;
+
+    @Test
+    @Tag("unit")
```
extract test tags and explain where they are used?

NicKIT01 replied:
Ich hab jetzt eine einheitliches Tagging mit custom tags eingeführt. Die kann man später in der pipeline auch verwenden und unit tests und integration tests getrennt voneinander ausführen

Jooshua replied:
Nice :)

Jooshua commented on backend/src/test/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/DirectoryRestAdapterTest.java:
```diff
@@ -0,0 +1,104 @@
+    DirectoryServicePort directoryService;
+
+    @Test
+    @DisplayName("POST /directory/ creates directory and returns 201")
```
No test tags?

Jooshua reviewed (approved):
Danke für die viele Arbeit und das Testen! :) 

Ich habe noch ein paar Kleinigkeiten angemerkt, die kann man aber ggf. auch in separaten Tickets lösen.

NicKIT01 committed:
#57: Refactored Delegator. Seperation of Concerns and use prefix defined by repository

NicKIT01 committed:
#57: DRY: Api constants for magic strings

NicKIT01 committed:
#57: Moved Exception Handler to global position

NicKIT01 committed:
#57: Unified testing strategy and tagging by using custom IntegrationTest and UnitTest JUnit5 Tags

NicKIT01 committed:
#57: Readded accidentally deleted library functionality, but now in hexagonal architecture.

NicKIT01 committed:
#57: Adjusted Frontend according to the new api of gates. Refactor to Single Source of Truth

NicKIT01 committed:
#57: Fixed cursor bug, when hovering above text files or library elements

NicKIT01 committed:
#57: Added Readme description for backend

NicKIT01 committed:
#57: Updated ARCHITECTURE.md

Jooshua commented on backend/src/test/resources/test-gates.json:
```diff
@@ -0,0 +1,16 @@
+  {
+    "name": "CNOT",
+    "symbol": "CNOT",
+    "type": "Classical",
```
Anmerkung: ein CNOT ist nicht classical

NicKIT01 replied:
upsi sind aber nur tests aber werde es noch abändern

NicKIT01 replied:
Ah ne, hatte das richtig so übernommen aus der json von Manuel. Da müssen wir dann im neuen Jahr nochmal über die Types gehen, deswegen lasse ich es erstmal so.

Jooshua reviewed (approved):
Habe noch einmal über die Kommentare / Changes drüber geschaut und passt alles von meiner Seite. Bei den Gates am Besten noch einmal mit @manuelcorcoles abstimmen?

simonvonroenn commented on backend/src/main/java/edu/kit/quak/core/library/model/Gate.java:
Wir müssen dann gucken, wie wir das dann mit der backend implementation vom circuit zusammenbringen. Weil da habe ich die gates ja auch nochmal definiert.

NicKIT01 replied:
Ich glaube ich würd dir hier widersprechen. Library sollte eine eigene Domain sein. Library könnte von Circuit aufgerufen werden aber ich würde Stammdaten (Library) und Zustand (Circuit) strikt voneinander trennen. Man will um die Library abzurufen ja nicht den Circuit abrufen. Das ist aber aktuell auch nur eine 1zu1 Übertragung von Manus vorhandener Funktionalität aus Development in die neue Architektur. Das kann man hinterher auf jeden Fall refactorn oder anders machen, ist aber gerade notwendig, damit Development mindestens auf dem gleichen Stand bleibt wie zuvor nach meinem Merge.

simonvonroenn commented on backend/src/main/java/edu/kit/quak/infrastructure/filesystem/out/db/jpa/DirectoryJpaAdapter.java:
```diff
@@ -0,0 +1,56 @@
+    public Directory save(Directory container) {
+        JpaDirectory jpaDirectory = directoryMapper.toJpaEntity(container);
+        // We need to set the parent of container manually because it is ignored by the mapping
+        // else we would lose the bidirectional behavior in the db
```
Warum kann man nicht einfach die parent id **nicht** ignorieren im mapping?

NicKIT01 replied:
Weil parentId auf parent gemappt werden soll. Es gibt keine parentId in der Datenbank.

simonvonroenn commented on backend/src/main/java/edu/kit/quak/infrastructure/library/out/json/JsonGateRepositoryAdapter.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    private record JsonGateDto(
+            @JsonProperty("name") String name,
+            @JsonProperty("type") String type,
+            @JsonProperty("description") String description,
+            @JsonProperty("qubitCount") int qubitCount,
+            @JsonProperty("symbol") Gate.SYMBOL symbol
+    ) {
+        Gate toDomain() {
+            return new Gate(name, type, description, qubitCount, symbol);
+        }
+    }
```
Warum ist das dto hier im adapter?

NicKIT01 replied:
Das ist ein DTO zwischen Adapter und JSON file. Ist privat und innerhalb des Adapters, weil nur der Adapter das braucht und sonst das nirgendwo gesehen werden muss. DTO bedeutet Data Transfer Object und bestimmt wie Daten ausgetauscht werden. Ist also nicht nur in Web/Rest Calls zu verwenden. Das ist vielmehr ein Entwurfsmuster

simonvonroenn commented on backend/src/test/java/edu/kit/quak/application/filesystem/delegator/FileElementContainerRepositoryDelegatorTest.java:
```diff
@@ -0,0 +1,88 @@
+@ExtendWith(MockitoExtension.class)
+class FileElementContainerRepositoryDelegatorTest {
+
+    @Mock
+    FileElementContainerRepositoryRegistry registry;
+
+    @Mock
+    FileElementContainerRepositoryPort<Project> projectRepo;
+
+    FileElementContainerRepositoryDelegator delegator;
+
+    @BeforeEach
+    void setup() {
+        delegator = new FileElementContainerRepositoryDelegator(registry);
+    }
```
Hier könnte man wieder mit `@Spy` und `@InjectMocks` arbeiten

NicKIT01 replied:
Warum? https://stackoverflow.com/questions/15052984/what-is-the-difference-between-mocking-and-spying-when-using-mockito & https://stackoverflow.com/questions/16467685/difference-between-mock-and-injectmocks Anders als bei den Mappern brauchen wir hier ja nicht die konkreten Instanzen, sondern wollen lediglich das Verhalten mocken.

simonvonroenn commented on backend/src/test/java/edu/kit/quak/infrastructure/filesystem/in/web/rest/mapper/FileElementDtoMapperTest.java:
```diff
@@ -0,0 +1,57 @@
+    // Configuration to pick up the MapStruct generated implementations
+    @Configuration
+    @ComponentScan(basePackageClasses = FileElementDtoMapper.class)
+    static class MapperTestConfig {}
```
Wer braucht das?

NicKIT01 replied:
Niemand habs gelöscht 👍🏻

simonvonroenn commented on backend/src/test/java/edu/kit/quak/infrastructure/library/out/json/JsonGateRepositoryAdapterTest.java:
Warum ist dieser Adapter-Test nur ein Unit-Test?

NicKIT01 replied:
Mein Gedanke war Integration Tests sind Tests, die einen gewissen Kontext hochfahren (Spring JPAData etc.). Ich glaube das ist hier etwas tricky und bewegt sich irgendwo in der Mitte zwischen integration und Unit. Die einzige Legitimation, das als Integration Test zu markieren ist für mich das verwenden der JSON file. Da diese aber statisch ist war es für mich ein Unit Test und kein Integration Test. Hab bei dem aber auch länger überlegt. Was wäre deine Begründung für/gegen Unit/Integration?

simonvonroenn replied:
Mein monkey brain hat nur Adapter-Test gelesen und dachte an Integration-Test. 
Aber wenn du dir was dabei gedacht hast, dann kann es gerne so bleiben. :)

simonvonroenn commented on frontend/src/views/library-view/GateLibrary.tsx:
```diff
@@ -1,39 +1,22 @@
                     key={`${gate.name}-${index}`}
                     id={gate.name}
-                    type={gate.symbol}
+                    type={gate.symbol} // Nutzung des Symbols für die Darstellung
```
Keine deutschen kommentare im code

simonvonroenn commented on frontend/src/views/library-view/GateList.tsx:
```diff
@@ -1,127 +1,120 @@
-      groups[type].push(gate);
-    }
+        for (const gate of gates) {
+            // Fallback, falls das Backend kein 'type' Feld liefert (Feldname ggf. anpassen!)
```
Keine deutschen kommentare im code + ich denke man kann hier erwarten, dass das backend einen type liefert, oder?

NicKIT01 replied:
ja hast recht steht ja so im DTO

simonvonroenn commented on frontend/src/views/project-manager-view/util/FileElement.tsx:
```diff
@@ -26,46 +32,51 @@ export interface File extends FileElement {
  * @param elements The elements to sort
  */
-export function sort(elements: FileElement[]) {
+export function sort(elements: FileElementDto[]) {
```
Ich bin mir ziemlich sicher, dass es eine library (evtl. sogar native) lösung gibt, die einen sorting algorithmus bereit stellt. Besser sowas dann nutzen.

NicKIT01 replied:
Ich habe lodash gefunden ist tatsächlich relativ clean jetzt, schau nochmal drüber! Vielen Dank für den Hinweis :) https://lodash.com

simonvonroenn commented on frontend/src/views/project-manager-view/util/FileElement.tsx:
```diff
@@ -26,46 +32,51 @@ export interface File extends FileElement {
  * Parses a given {@link FileElement} into a {@link JSX.Element}
  * @param object The object to parse
  */
-export function getElementForFileElement(object: FileElement) {
+
+export function getElementForFileElement(object: FileElementDto) {
     if (object.type === "file") {
-        return (<File {...object as File} key = {object.id}/>)
+        return (<File {...object as unknown as FileDetailsResponse} key={object.id}/>)
     } else if (object.type === "directory") {
-        return (<Directory {...object as Directory} key = {object.id}/>)
+        return (<Directory {...object as unknown as DirectoryContentsResponse} key={object.id}/>)
     } else {
-        return (<Project {...object}/>)
+        return (<Project {...object as unknown as ProjectContentsResponse} key={object.id}/>) // ID als Key ist wichtig!
     }
```
Diese Methode evtl. nach src/api/mapper auslagern?

NicKIT01 replied:
Hmm ja, da hast du recht. Das muss ich mir morgen oder am Montag nochmal in Ruhe angucken. Eigentlich brauchen wir auch die interfaces FileElement etc. nicht mehr unbedingt.  Ich gehe da nochmal drüber!

NicKIT01 replied:
Werde es wahrscheinlich erst nach dem Wochenende bearbeiten

NicKIT01 replied:
Ich glaub das ist was, was ich in einem gesonderten Refactoring machen würde. Das wird relativ groß. Aktuell ist das ja keine richtige Mapper Methode sondern eher ein Discriminator oder eine Art UIFactory. Würde gerne richtiges Mapping integrieren und die aktuell genutzten shadow interfaces erweitern/ändern. Zusätzlich würde ich gerne Single Source of Truth einbauen für Project-manager-view, um das refactoring zu erleichtern, falls sich mal was an der UI ändern sollte. Ich erstelle dazu ein eigenes Ticket :)

NicKIT01 replied:
#82

simonvonroenn commented on frontend/src/views/text-editor-view/QLPEditor.tsx:
```diff
@@ -35,14 +36,40 @@ function QLPEditor({ file }: { file: File | undefined }) {
+            console.error("Failed to encode content", e);
+            throw e;
+        }
+    }
```
Keine deutschen Kommentare! + ich denke auch hier gibt es eine encode library, die das automatisch macht.

NicKIT01 replied:
Ja gibt ne base64 library👍🏻

simonvonroenn reviewed (changes requested):
Generelle Anmerkung: Evtl. sollten wir uns mit Furkan nochmal absprechen bzgl. namings (e.g. RestAdapter vs RestController).

NicKIT01 committed:
#57: Delete useless configuration

NicKIT01 committed:
#57: Removing german comments

NicKIT01 committed:
#57: Refactored sort algorithm using lodash

NicKIT01 committed:
#57: Refactored encoding content using js-base64

simonvonroenn commented on frontend/src/views/text-editor-view/QLPEditor.tsx:
```diff
@@ -35,14 +38,27 @@ function QLPEditor({ file }: { file: File | undefined }) {
+    function encodeContent(str: string): string {
+        return Base64.encode(str);
+    }
+
```
NIT: Kannst dann auch noch die function auflösen, wenn die eh nur ein Einzeiler ist und auch nur an einer Stelle benutzt wird.

NicKIT01 committed:
#57: Inline only function call

simonvonroenn commented on frontend/src/views/text-editor-view/QLPEditor.tsx:
```diff
@@ -109,20 +122,24 @@ function QLPEditor({ file }: { file: File | undefined }) {
+                } catch (e) {
+                    console.error("Failed to decode file content", e);
+                    return ""; // oder Fehler werfen
+                }
```
Hier dann auch noch das ersetzen mit Base64.decode()

NicKIT01 committed:
#57: Introduced base64 encoding and better error handling
